types/src/front/acl.ts

```diff
@@ -1,29 +1,84 @@
 import { ModelId } from "../shared/model_id";
+import { GroupType } from "./groups";
+import { LightWorkspaceType, RoleType } from "./user";
 
 // Supported permissions
-export const SUPPORTED_PERMISSIONS = ["read", "write"] as const;
+export const SUPPORTED_PERMISSIONS = ["admin", "read", "write"] as const;
 
 export type Permission = (typeof SUPPORTED_PERMISSIONS)[number];
 
-// Access Control Entry
-export type ACEType = {
-  groupId: ModelId;
+export type GroupPermission = {
+  id: ModelId;
   permissions: Permission[];
 };
 
-// Access Control List
-export type ACLType = {
-  aclEntries: Array<ACEType>;
+export type RolePermission = {
+  name: RoleType;
+  permissions: Permission[];
+};
+
+export type GroupOnlyACL = {
+  groups: GroupPermission[];
+  roles?: never[];
 };
 
-export function groupHasPermission(
+export type RoleBasedACL = {
+  groups: GroupPermission[];
+  roles: RolePermission[];
+  workspaceId: ModelId; // Required when roles are defined.
+};
+
+function hasRoles(acl: ACLType): acl is RoleBasedACL {
+  return "roles" in acl;
+}
+
+export type ACLType = GroupOnlyACL | RoleBasedACL;
```

Comment:
What's the point of having 2 sub types ?

Reply:
There are some places where we don't have a workspace, in those places we don't want to leverage the role based access. Creating an exclusion type allow us to not having to pass workspaces in all those places.

Reply:
GroupAndRoleACL then ?